[C++] Object API: document custom string type requirements, implement better custom string type constructor alternative for Unpack() and fix bug with vector of custom string types in Pack().

llongi opened the pull request:
When using the C++ Object API, it is possible to define an own string type via --cpp-str-type or the equivalent attribute. The documented requirements for this are that the type must implement T::c_str() and T::length().
Currently generated code also requires T::empty() due to a generated check in Pack(). empty() should be equivalent to length() == 0, and is thus a functionally equivalent replacement, that does not pose additional requirements on the type T to implement empty().
The documentation could have also been changed to simply require T::empty(), but I find this a cleaner approach that does not change the theoretically documented requirements.

aardappel commented:
Hmm, I think just adding `empty()` to the requirements would have been better, it is more canonical to use `empty()`, and changing all the code to look clumsier to save string classes implementing `empty()` does not seem like a good tradeoff.

llongi commented:
That's the other solution and is fine by me too. I'll update the PR tomorrow to use and document the empty requirement properly. I'll also document that the custom class does need to have a constructor or assignment operator taking std::string, given that FB->str() is always used to construct it in Unpack(). That is somewhat unfortunate, as it means you cannot make a zero-copy-construction version of your own string class, and you always incur at least two copies (once to build the std::string, and once to copy the memory to your own implementation, since you cannot safely move the underlying storage from std::string). I was thinking of proposing to change that ->str() call to be ->string_view() (if HAS_STRING_VIEW), as that would, at least for recent C++, allow all problems above to be nicely solved. But that would break the current API, as any custom string class would, right now, implement a constructor from std::string, but not necessarily from string_view, and since there's no automatic string_view to std::string conversion, implementors would have to take action.
Thoughts on this? It's not a crucial optimization or use case for me, but I did take note of the above possible limitations.

aardappel commented:
I agree that is unfortunate, and indeed `string_view` is the natural choice for this. In fact, I'd prefer `str()` not to exist at all, since a forced copy is so not FlatBuffers.

But indeed we cannot break that API at this point, so it would need.. yet another option or attribute. Not sure if you feel that is worth the trouble.

llongi commented:
I could add an option to do that, but what would it be concretely?
A) custom_str(e->string_view()) is the cleanest approach, but only C++14/17 and up. Could be ifdef HAS_STRING_VIEW, but still needs an alternative for older language standards.
B) custom_str(e->c_str(), e->length()) is basically the portable version.
C) custom_str(e), where e is a flatbuffers::String directly, the most flexible approach.

Personally I would go for C, as that gives the maximum flexibility: the implementor can then decide which way to access the flatbuffers::String is appropriate for his use case.

aardappel commented:
I think B) is the best, since most custom strings will already have such a constructor, and it is the most flexible.
C) is problematic because it may require for a your custom string header to depend on flatbuffers.h

llongi commented:
True, didn't think of the dependency on flatbuffers.h. Let's go for B then.
Suggestions for the switch/attribute name? --cpp-str-flex-ctor, --cpp-str-no-string-ctor, ...

llongi commented:
There it is.
- Updated all the documentation to mention empty() and construct from std::string to be a requirement.
- Added --cpp-str-flex-ctor switch and cpp_str_flex_ctor option attribute to use custom_string_constructor(e->c_str(), e->size()) instead.
- Fixed bug with vector of custom string types (present in current code). CreateVectorOfStrings() was always used with string types, but required std::string, which fails to compile if the type is a custom string type instead. Switching to CreateVector() with function variant does the same operation with the same performance, but works with all types of strings.

llongi commented:
Also just a quick note: the first commit is just a clang-format run on idl_gen_cpp.cpp., to ensure any subsequent changes were properly formatted. I hope that's not a problem, at least it helps in having consistent formatting in the file again.

aardappel commented on docs/source/CppUsage.md:
```diff
@@ -255,14 +255,24 @@ you, so you'll have to manage their lifecycles manually.  To reference the
+so that the custom string type is constructed by passing the pointer and
+length of the FlatBuffers String. The custom string class will require a
+constructor in the following format: custom_str_class(const char *, uint32_t).
+Please note that the character array is not guaranteed to be NUL terminated,
```
NULL

llongi replied:
Fixed.

aardappel commented on docs/source/CppUsage.md:
```diff
@@ -255,14 +255,24 @@ you, so you'll have to manage their lifecycles manually.  To reference the
+per field attribute `cpp_str_flex_ctor` can be used to change this behavior,
+so that the custom string type is constructed by passing the pointer and
+length of the FlatBuffers String. The custom string class will require a
+constructor in the following format: custom_str_class(const char *, uint32_t).
```
probably makes more sense to require a `size_t` argument

llongi replied:
True, FB always has a uint32_t for length(), so that's why I suggested it. Using size_t also works and is probably more natural and what people will have implemented already.

aardappel commented on src/idl_gen_cpp.cpp:
```diff
@@ -494,8 +495,8 @@ class CppGenerator : public BaseGenerator {
-    static const char * const ctypename[] = {
-    // clang-format off
+    static const char *const ctypename[] = {
+// clang-format off
```
Not sure why it is doing this, but this change is not desirable. It may need a blank line somewhere to work, I seem to recall.

llongi replied:
Yeah, clang-format seems to be confused about how to properly align the comments themselves while inside a block. Just putting the on/off comments outside the block makes everything work consistently, fixed in 117e3b0679209f2aa55cbee18c4036e7da4bd4b3.

aardappel commented on src/idl_gen_cpp.cpp:
```diff
@@ -577,12 +578,17 @@ class CppGenerator : public BaseGenerator {
+  bool FlexibleStringConstructor(const FieldDef *field) {
+    auto attr = field ? (field->attributes.Lookup("cpp_str_flex_ctor") != nullptr) : false;
+    auto ret = attr ? attr : parser_.opts.cpp_object_api_string_flexible_constructor;
+    return (ret && NativeString(field) != "std::string"); // Only for custom string types.
```
remove `()`

llongi replied:
Done.

aardappel commented on src/idl_gen_cpp.cpp:
```diff
@@ -577,12 +578,17 @@ class CppGenerator : public BaseGenerator {
     if (ptr_type != "naked") {
-      return (ptr_type != "default_ptr_type" ? ptr_type :
-              parser_.opts.cpp_object_api_pointer_type) + "<" + type + ">";
+      return (ptr_type != "default_ptr_type" ? ptr_type : parser_.opts.cpp_object_api_pointer_type) + "<" + type + ">";
```
make sire all lines are <80 columns

llongi replied:
In the first clang-format run I did, I disabled the enforcement of 80 chars lines, mostly because as you can see from the new commit 4791923806965637d5b13f7003329bfbb2fdf18b it formats a lot of code. But now it's fully compliant with the provided clang-format file, which it wasn't before. It's a good thing, I just wasn't sure if you'd accept this kind of changes in a PR.

aardappel replied:
Typically prefer `git clang-format` to limit changes to the new code, yes.

aardappel commented on src/idl_gen_cpp.cpp:
```diff
@@ -2332,7 +2341,13 @@ class CppGenerator : public BaseGenerator {
-            code += "_fbb.CreateVectorOfStrings(" + value + ")";
+            // Use by-function serialization to emulate CreateVectorOfStrings();
+            // this works also with non-std strings.
+            code += "_fbb.CreateVector<flatbuffers::Offset<flatbuffers::String>> ";
```
This is not great, as it creates more code for all users, even though it is needed for a tiny minority. Would be better to leave the original in there for most cases.

llongi replied:
Okay, now if type is std::string the normal CreateVectorOfStrings() is used, in all other cases the function.

aardappel commented:
Thanks!

llongi commented:
Thank you! And I didn't know about git clang-format, looks like a useful tool.
But somehow I can't find the code showing up in google/flatbuffers:master, even if here it says you merged it to it? Is there some delay or weird git at work?

aardappel commented:
Apologies, not sure what went wrong there. I couldn't find it either, so I manually squashed and merged your PR https://github.com/google/flatbuffers/commit/eb2a81f73db9893fbe2eec53662fa7f9358e54a8

This does mean it does not show you as the author of the commit, though it does refer to you in the commit message. Sorry about that.

llongi commented:
That's perfectly fine. Thank you!
